### PR TITLE
[Issues 163 + 128 ] build kraken via golang docker container.

### LINKED ISSUE #163: release tarballs for kraken cli tool are now getting created with a capital "K"
This seems to be new. The following list is not exhaustive and I surmise that given the sampling below the issue is systemic in the build now.

`$ for f in {k,g}*gz;do echo $f;tar tvfz $f;done`
```
k2cli_1.1.0_darwin_amd64.tar.gz
-rwxr-xr-x root/root  11862592 2017-09-15 13:35 Kraken
k2cli-release_1.0.9_linux_amd64.tar.gz . # <<--- old build which is lower case
-rwxr-xr-x root/root  11843443 2017-08-25 11:11 k2cli
kraken-rel_1.1.0_linux_amd64.tar.gz
-rwxr-xr-x root/root  11871460 2017-09-15 13:35 Kraken
gox_1.1.0_darwin_amd64.tar.gz
-rwxr-xr-x root/root   3121424 2017-09-15 13:35 Kraken
```

### PR DESCRIPTION
**Pull request by alejandroEsc**
**What this PR does / why we need it**: Building kraken via golang container.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #163 and #128

**Special notes for your reviewer**: Testing in pipeline will be verified that nothing breaks. Removed godeps as we are sing go vendor.

**Release note**:

```
Kraken is now built inside golang container, version 1.8.3. Build version for an OS version is defined via GOOS=' (default is linux) environment variable.
```

### CODE REVIEW

**Comment by alejandroEsc** [comment]
ok it builds fine, which is to be expected, next I will try to do a new section dedicated in making sure the builds and compilation process works.

File: Jenkinsfile [hunk 1]
```diff
@@ -6,15 +6,21 @@ release_version        = "${env.RELEASE_VERSION}"
...
-podTemplate(label: 'kraken', containers: [
-    containerTemplate(name: 'jnlp', image: 'quay.io/samsung_cnct/custom-jnlp:0.1', args: '${computer.jnlpmac} ${computer.name}'),
-    containerTemplate(name: 'golang', image: 'quay.io/samsung_cnct/kraken-gobuild:1.8.3', ttyEnabled: true, command: 'cat', alwaysPullImage: true),
-    containerTemplate(name: 'kraken-tools', image: 'quay.io/samsung_cnct/k2-tools:latest', ttyEnabled: true, command: 'cat', alwaysPullImage: true, resourceRequestMemory: '1Gi', resourceLimitMemory: '1Gi'),
-    ], volumes: [
+podTemplate(
+    label: 'kraken',
+    containers: [
+        containerTemplate(name: 'jnlp', image: 'quay.io/samsung_cnct/custom-jnlp:0.1', args: '${computer.jnlpmac} ${computer.name}'),
+        containerTemplate(name: 'golang', image: 'quay.io/samsung_cnct/kraken-gobuild:1.8.3', ttyEnabled: true, command: 'cat', alwaysPullImage: true),
```
Comment (mikeln) [inline Jenkinsfile]: Why another image?  we have samsung_cnct/goglide:1.8.3  (or 1.9.0) for this, or, probably better the golang:1.8.3 container image.

File: Jenkinsfile [hunk 2]
```diff
@@ -29,7 +35,6 @@ podTemplate(label: 'kraken', containers: [
                 withEnv(["GOPATH=${WORKSPACE}/go/"]) {
```
Comment (mikeln) [inline Jenkinsfile]: This doesn't do what you think.     This is setting a Jenkins groovy variable GOPATH, not a shell env variable that you need.

File: Jenkinsfile [hunk 3]
```diff
@@ -40,75 +45,89 @@ podTemplate(label: 'kraken', containers: [
...
-                    kubesh "go/src/github.com/samsung-cnct/kraken/kraken generate --provider aws /var/lib/docker/scratch/kraken-${env.JOB_BASE_NAME}-${env.BUILD_ID}/aws/config.yaml"
-                    kubesh "go/src/github.com/samsung-cnct/kraken/build-scripts/update-generated-config.sh /var/lib/docker/scratch/kraken-${env.JOB_BASE_NAME}-${env.BUILD_ID}/aws/config.yaml kca${env.JOB_BASE_NAME}-${env.BUILD_ID} /var/lib/docker/scratch/kraken-${env.JOB_BASE_NAME}-${env.BUILD_ID}"
-                    kubesh "go/src/github.com/samsung-cnct/kraken/kraken generate --provider gke /var/lib/docker/scratch/kraken-${env.JOB_BASE_NAME}-${env.BUILD_ID}/gke/config.yaml"
-                    kubesh "go/src/github.com/samsung-cnct/kraken/build-scripts/update-generated-config.sh /var/lib/docker/scratch/kraken-${env.JOB_BASE_NAME}-${env.BUILD_ID}/gke/config.yaml kcg${env.JOB_BASE_NAME}-${env.BUILD_ID} /var/lib/docker/scratch/kraken-${env.JOB_BASE_NAME}-${env.BUILD_ID}"
+            customContainer('install-make') {
+                sh "apt-get update && apt-get -y install make && cp /usr/bin/make /out/make"
```
Comment (mikeln) [inline Jenkinsfile]: should use kubesh instead of sh to work around/avoid jenkins issue.

File: Jenkinsfile [hunk 4]
```diff
@@ -40,75 +45,89 @@ podTemplate(label: 'kraken', containers: [
...
+            }
 
+            withEnv(["GOPATH=${WORKSPACE}/go/"]) {
+                stage('Make: build, install, compile') {
```
Comment (mikeln) [inline Jenkinsfile]: Again, this sets the groovy var GOPATH, not the bash environment var.

File: Jenkinsfile [hunk 5]
```diff
@@ -40,75 +45,89 @@ podTemplate(label: 'kraken', containers: [
...
+                            kubesh "rm -rf /var/lib/docker/scratch/kraken-${env.JOB_BASE_NAME}-${env.BUILD_ID}/gke"
+                        }
+                    )
+                }
+            }
+        }
+
+        if (git_branch.contains(publish_branch) && git_uri.contains(github_org)) {
+            customContainer('golang') {
+                withEnv(["GOPATH=${WORKSPACE}/go/"]) {
```
Comment (mikeln) [inline Jenkinsfile]: see previous comments.

File: Jenkinsfile [hunk 6]
```diff
@@ -40,75 +45,89 @@ podTemplate(label: 'kraken', containers: [
...
     }
-  }
 
-  sh(command)
+    sh(command)
 }
 
 def kubectl(command) {
-  "kubectl exec -i ${env.HOSTNAME} -c ${env.CONTAINER_NAME} -- /bin/sh -c 'cd ${env.WORKSPACE} && export GOPATH=${env.GOPATH} && ${command}'"
+    "kubectl exec -i ${env.HOSTNAME} -c ${env.CONTAINER_NAME} -- /bin/sh -c 'cd ${env.WORKSPACE} && export GOPATH=${env.GOPATH} && ${command}'"
```
Comment (mikeln) [inline Jenkinsfile]: Ok, this is my answer:
`"kubectl exec -i ${env.HOSTNAME} -c ${env.CONTAINER_NAME} -- /bin/sh -c 'cd ${pwd()} && ${command}'"`

This way the command will run at the location your pipeline step is currently at. (vs always throwing you back to the base WORKSPACE location.)  

But I see that this is how you are getting the groovy GOPATH into the bash env.  Ignore my above comments, if you don't mind always setting that var.

File: Makefile [hunk 7]
```diff
@@ -1,39 +1,51 @@
...
 COMMIT    := $(shell git rev-parse HEAD)
 REL_BRANCH := "$$(git rev-parse --abbrev-ref HEAD)"
-godep=GOPATH=$(shell godep path):${GOPATH}
-
+PROJECT_PATH := "/go/src/github.com/samsung-cnct/kraken/"
+GO_CONTAINER := "quay.io/samsung_cnct/kraken-gobuild"
+GO_VERSION := 1.8.3
+GO_DOCKER_SRC_VOLUME := ${GOPATH}/src:/go/src
+GO_DOCKER_BIN_VOLUME := ${GOPATH}/bin:/go/bin
+GOOS?=linux
```
Comment (mikeln) [inline Makefile]: Ok, I'm not in favor of this method.  Running Docker directly out of the building Makefile.
You should look at golang-tools/example-project.

The issue here is that you assume that the user of the Makefile has docker running.  There are several issues with this.
1) It clouds the intent of the Makefile.  The actual build, is intertwined with a build environment. 
2) It does not handle detecting of being able to actually run...due to missing build env.

The build Makefile should stand on its own, and assume a valid golang env for building. 
Outside of the Makefile you can wrap a shell script to validate/start/ensure that a build environment is correct.  e.g. that Docker is installed and running.   Then you also include the commands to run the makefile inside a golang contianer there.  This allows the developer to either 1) have their own golang env setup and run the makefile directly by hand, or 2) only have docker installed, and run the script to build via running the Makefile inside the golang container.
Also, this separation works much better with Jenkins, as Jenkins can then fire up a golang container and run the Makefile directly.  i.e.  Jenkins worries about setting up the correct environment for the build.
Like I said see golang-tools/example-project.

File: Jenkinsfile (hunk 2 above)
Comment (alejandroEsc) [reply Jenkinsfile]: I didnt write this. But yeah we should talk about it.

**Comment by alejandroEsc** [comment]
closing this PR, some details here need to be discussed more and we should decide on a convention for this project that suites everyones needs.

**Pull request closed without merging**